contracts/characters.sol

```diff
@@ -244,44 +251,46 @@ contract Characters is Initializable, ERC721Upgradeable, AccessControlUpgradeabl
-            garrison.updateOnBurn(burnOwner, burnId);
-        }
-        _burn(burnId);
-        bonusPower[targetCharId] = burnPower.add(bonusPower[targetCharId]);
+    function burnIntoCharacter(uint256[] calldata burnIds, uint256 targetCharId) external restricted {
+        uint256 burnPower = 0;
+        for(uint i = 0; i < burnIds.length; i++) {
+            burnPower += nftVars[NFTVAR_BONUS_POWER][burnIds[i]].add(getPowerAtLevel(tokens[burnIds[i]].level));
+            address burnOwner = ownerOf(burnIds[i]);
+            if(burnOwner == address(garrison)) {
+                burnOwner = garrison.characterOwner(burnIds[i]);
+                garrison.updateOnBurn(burnOwner, burnIds[i]);
+            }
+            _burn(burnIds[i]);
 
-        emit Burned(
-            burnOwner,
-            burnId
-        );
+            emit Burned(
+                burnOwner,
+                burnIds[i]
+            );
+        }
+        require(uint(4).mul(getPowerAtLevel(tokens[targetCharId].level)) >= getTotalPower(targetCharId).add(burnPower), "Power limit");
+        nftVars[NFTVAR_BONUS_POWER][targetCharId] = burnPower.add(nftVars[NFTVAR_BONUS_POWER][targetCharId]);
     }
 
-    function burnIntoSoul(uint256 burnId, address soulRecipient) external restricted {
-        uint256 soulAmount = uint256(getPowerAtLevel(tokens[burnId].level)).add(bonusPower[burnId]).div(10);
-        address burnOwner = ownerOf(burnId);
-        if(burnOwner == address(garrison)) {
-            garrison.updateOnBurn(burnOwner, burnId);
-        }
-        _burn(burnId);
-        soulSupply[soulRecipient] = soulSupply[soulRecipient].add(soulAmount);
+    function burnIntoSoul(uint256[] calldata burnIds) external restricted {
+        for(uint i = 0; i < burnIds.length; i++) {
+            address burnOwner = ownerOf(burnIds[i]);
+            if(burnOwner == address(garrison)) {
+                burnOwner = garrison.characterOwner(burnIds[i]);
+                garrison.updateOnBurn(burnOwner, burnIds[i]);
+            }
+            _burn(burnIds[i]);
 
-        emit Burned(
-            soulRecipient,
-            burnId
-        );
+            emit Burned(
+                burnOwner,
+                burnIds[i]
+            );
+        }
     }
 
-    function upgradeWithSoul(uint256 targetCharId, uint256 soulAmount) public {
-        require(soulSupply[msg.sender] >= soulAmount, 'Not enough soul');
+    function upgradeWithSoul(uint256 targetCharId, uint256 soulAmount) external restricted {
         uint256 burnPower = soulAmount.mul(10);
         require(uint(4).mul(getPowerAtLevel(tokens[targetCharId].level)) >= getTotalPower(targetCharId).add(burnPower), "Power limit");
-        soulSupply[msg.sender] = soulSupply[msg.sender].sub(soulAmount);
-        bonusPower[targetCharId] = burnPower.add(bonusPower[targetCharId]);
+        nftVars[NFTVAR_BONUS_POWER][targetCharId] = burnPower.add(nftVars[NFTVAR_BONUS_POWER][targetCharId]);
```

Comment:
The order of keys is flipped for nftVars here.
It's supposed to be targetCharId first, then nft var second

Reply:
actually in a couple other places too (line 309, line 271, 257)